proof/invariant-abstract/AARCH64/ArchFPU_AI.thy

```diff
@@ -0,0 +1,381 @@
+(*
+ * Copyright 2025, Proofcraft Pty Ltd
+ *
+ * SPDX-License-Identifier: GPL-2.0-only
+ *)
+
+theory ArchFPU_AI
+imports ArchDetSchedSchedule_AI
+begin
+
+context Arch begin arch_global_naming
+
+section \<open>cur_fpu_in_cur_domain\<close>
+
+definition cur_fpu_in_cur_domain :: "det_ext state \<Rightarrow> bool" where
+  "cur_fpu_in_cur_domain s \<equiv> none_top (\<lambda>t. in_cur_domain t s) (arm_current_fpu_owner (arch_state s))"
+
+\<comment> \<open>FIXME: move\<close>
+lemma none_top_post_wp:
+  "\<lbrakk>\<And>x. \<lbrace>P x\<rbrace> f \<lbrace>\<lambda>_. Q x\<rbrace>\<rbrakk>
+   \<Longrightarrow> \<lbrace>\<lambda>s. none_top (\<lambda>x. P x s) y\<rbrace> f \<lbrace>\<lambda>_ s. none_top (\<lambda>x. Q x s) y\<rbrace>"
```

Comment:
Nice. Is there are good theory to put this one in? The wp theories probably don't include none_top yet and vice versa. Maybe somewhere else in Lib then.

Reply:
I guess `lib/NonDetMonadLemmaBucket`? I'm not sure if that's a file we actually want to keep around but for now it seems like the most appropriate place.

Reply:
Yes, that one would work. I don't think that file will go away completely, but maybe it can be renamed at some point to serve as a point for things like these that need a larger context.